src/qibolab/instruments/zhinst.py

```diff
@@ -0,0 +1,1041 @@
+                                integration_kernel=weight,
+                                integration_kernel_parameters=None,
+                                integration_length=None,
+                                measure_signal=f"measure{qubit.name}",
+                                measure_pulse=pulse.zhpulse,
+                                measure_pulse_length=round(pulse.pulse.duration * 1e-9, 9),
+                                measure_pulse_parameters=measure_pulse_parameters,
+                                measure_pulse_amplitude=None,
+                                acquire_delay=self.time_of_flight * 1e-9,
+                                reset_delay=relaxation_time * 1e-9,
+                            )
+                            i += 1
+
+        if optimized:
+            print("I'm using optimized IW")
+        else:
+            print("I'm using dumb IW")
+
+    def fast_reset(self, exp, qubits, fast_reset):
+        """
+        Conditional fast reset after readout - small delay for signal processing
+        This is a very naive approach that can be improved by repeating this step until
+        we reach non fast reset fidelity
+        https://quantum-computing.ibm.com/lab/docs/iql/manage/systems/reset/backend_reset
+        """
+        print("Im fast resetting")
+        for qubit_name in self.sequence_qibo.qubits:
+            qubit = qubits[qubit_name]
+            if not qubit.flux_coupler:
+                with exp.section(uid=f"fast_reset{qubit.name}", play_after=f"sequence_measure{qubit.name}"):
+                    with exp.match_local(handle=f"sequence{qubit.name}"):
+                        with exp.case(state=0):
+                            pass
+                        with exp.case(state=1):
+                            exp.play(signal=f"drive{qubit.name}", pulse=ZhPulse(fast_reset[qubit.name]).zhpulse)
+
+    def sweep(self, qubits, sequence, options, *sweepers):
+        """Play pulse and sweepers sequence"""
+
+        self.signal_map = {}
+
+        sweepers = list(sweepers)
+
+        dimensions = []
+        if options.averaging_mode is AveragingMode.SINGLESHOT:
+            dimensions = [options.nshots]
+
+        for sweeper in sweepers:
+            dimensions.append(len(sweeper.values))
+
+        # Re-arranging sweepers based on hardware limitations
+        # FIXME: Punchout and frequency case
+        rearranging_axes = [[], []]
+        if len(sweepers) == 2:
+            if sweepers[1].parameter is Parameter.frequency:
+                if (
+                    not sweepers[0].parameter is Parameter.amplitude
+                    and sweepers[0].pulses.type is not PulseType.READOUT
+                ):
+                    rearranging_axes[0] += [sweepers.index(sweepers[1])]
+                    rearranging_axes[1] += [0]
+                    sweepers.remove(sweepers[1])
+                    sweepers.insert(0, sweepers[1])
+                    warnings.warn("Sweepers were reordered")
+
```

Comment:
I understand that this was done because the frequency must always be swept first in Zurich. If that is the case, it would be true regardless the number of `sweepers` used, thus `len(sweepers)==2` will not fix the problem. We do not have this use case for now indeed.

Reply:
we don't use 3 sweepers not so it may be not as simple as just frequency on top as it hapenned with Amplitude Readout.